peps/pep-0817.rst

```diff
@@ -868,56 +944,60 @@ distinguishable from regular variants.
-desirable to avoid plugin use at install time for the second class of
-variants, where it is not strictly necessary.
-
-To account for this, the specification proposes two classes of
-providers:
-
-a. Install-time providers that require executing code to determine
-   the wheel compatibility, and therefore satisfy the variants needing
-   that.
-
-b. Ahead-of-Time (AoT) providers that entirely rely on static metadata
-   for the purpose of wheel selection, and that can be used for the
-   other class of variants to avoid the additional problems posed
-   by install-time providers.
-
-For user convenience and better consistency across different packages,
-AoT providers can also feature plugin packages, but they are only used
-at build time (where executing code from additional packages is already
-a risk) to verify the variant properties specified by the maintainer
-and automatically generate the static metadata.
-
-Both kinds of provider plugins expose the same API. However, an AoT
-provider must always consider all valid variant properties supported,
-and it must always return the same ordered list of supported properties
-irrespective of the user system. All AoT providers can be technically
-used as install-time providers, but not the other way around.
+The variant wheel metadata specifies what providers are used for its
+properties. Providers serve twofold purpose:
+
+a. at install time: determining which variant wheels are compatible with
+   the user's system, and which of them constitutes the best choice, and
+
+b. at build time: determining which variant properties are valid for
+   building a wheel.
+
+The specification proposes two kinds of providers: install-time
+providers and Ahead-of-Time providers.
+
+Install-time providers are implemented either as Python packages that
+need to be installed and ran to query them, or vendored or reimplemented
+in the tools. They are used when user systems need to be queried to
+determine wheel compatibility, for example for variants utilizing GPUs
+or requiring CPU instruction sets beyond what platform tags provide.
+Installing third-party packages involves security risks highlighted in
+the `security implications`_ section, and the proposed mitigations incur
+a cost on installer implementations.
+
+Ahead-of-Time providers are implemented as static metadata embedded in
+the wheel. They are used when particular variant properties are always
+compatible with the user's system (provided that a wheel using them has
+been built successfully). However, the metadata indicates which
+properties are preferred. For example, AoT providers can be used to
+provide choice between builds against different BLAS / LAPACK providers,
+or to provide debug builds of packages. Since they do not require
+running code external to the installer, they do not pose the problems
+faced by install-time providers, and can be used more liberally.
+
+AoT providers are permitted to feature plugin packages. If that is the
```

Comment:
This first sentence seems confusing. It raises the question of what the alternative is, since they do need to be implemented somewhere and presumably that's in a plugin package.

Reply:
Actually, the previous paragraph states that they are "implemented as static metadata embedded in            the wheel".